### PR TITLE
Tracker Changes

### PR DESCRIPTION
**Pull request by kloon**
This PR introduces some changes to the tracker data collection.

### CODE REVIEW

File: includes/class-wc-tracker.php [hunk 1]
```diff
@@ -391,11 +435,80 @@ private static function get_all_template_overrides() {
...
+			'gross' => $gross_total,
+		);
+	}
+
+	/**
+	 * Get last order date
+	 *
+	 * @return string
+	 */
+	private static function get_last_order_date() {
```
Comment (mikejolley) [inline includes/class-wc-tracker.php]: Why not do this with direct SQL like `get_order_totals`? Since it's internal we can - it will be faster as you're only querying the posts table.

Comment (mikejolley) [reply includes/class-wc-tracker.php]: That would also allow you to get both dates (first/last) at the same time using min/max.

**Review by mikejolley — CHANGES_REQUESTED** [review]
Added some feedback.

This is on hold pending the notice. Can you rebase to remove those unrelated commit messages? Or we can squash. Since this has changed quite a bit from first draft :)

**Comment by kloon** [comment]
@mikejolley squash and merge is fine.

File: includes/class-wc-tracker.php [hunk 2]
```diff
@@ -391,11 +432,85 @@ private static function get_all_template_overrides() {
...
+
+		return $min_max;
+	}
+
+	/**
+	 * Get first order date.
+	 *
+	 * @return string
+	 */
+	private static function get_first_order_date() {
```
Comment (mikejolley) [inline includes/class-wc-tracker.php]: Is this still needed?

**Review by mikejolley — CHANGES_REQUESTED** [review]

File: includes/class-wc-tracker.php (hunk 2 above)
Comment (kloon) [reply includes/class-wc-tracker.php]: Missed it, removed now.

File: includes/class-wc-tracker.php [hunk 3]
```diff
@@ -391,11 +432,64 @@ private static function get_all_template_overrides() {
...
+	 * @return string
+	 */
+	private static function get_order_dates() {
+		global $wpdb;
+
+		$min_max = $wpdb->get_row( "
+			SELECT
+				MIN( post_date_gmt ) as 'first', MAX( post_date_gmt ) as 'last'
+			FROM {$wpdb->prefix}posts
+			WHERE post_type = 'shop_order'
```
Comment (mikejolley) [inline includes/class-wc-tracker.php]: Do we also want a status check here or is it not important?

File: includes/class-wc-tracker.php [hunk 4]
```diff
@@ -5,21 +5,25 @@
  * The WooCommerce tracker class adds functionality to track WooCommerce usage based on if the customer opted in.
  * No personal information is tracked, only general WooCommerce settings, general product, order and user counts and admin email for discount code.
  *
- * @class 		WC_Tracker
- * @version		2.3.0
- * @package		WooCommerce/Classes
- * @category	Class
- * @author 		WooThemes
+ * @class WC_Tracker
+ * @version 2.3.0
```
Comment (mikejolley) [inline includes/class-wc-tracker.php]: May as well drop this if we're not versioning the changes.

File: includes/class-wc-tracker.php [hunk 5]
```diff
@@ -5,21 +5,25 @@
...
  */
 
 if ( ! defined( 'ABSPATH' ) ) {
 	exit;
 }
 
+/**
+ * WooCommerce Tracker Class
+ *
+ * @version 2.3.0
```
Comment (mikejolley) [inline includes/class-wc-tracker.php]: As above.

File: includes/class-wc-tracker.php [hunk 6]
```diff
@@ -5,21 +5,25 @@
...
- * @version		2.3.0
- * @package		WooCommerce/Classes
- * @category	Class
- * @author 		WooThemes
+ * @class WC_Tracker
+ * @version 2.3.0
+ * @package WooCommerce/Classes
  */
 
 if ( ! defined( 'ABSPATH' ) ) {
```
Comment (mikejolley) [inline includes/class-wc-tracker.php]: Switch the our 1 line defined code :)

File: includes/class-wc-tracker.php [hunk 7]
```diff
@@ -34,10 +38,10 @@ public static function init() {
 	/**
 	 * Decide whether to send tracking data or not.
 	 *
-	 * @param boolean $override
+	 * @param boolean $override Should override?.
 	 */
 	public static function send_tracking_data( $override = false ) {
-		// Don't trigger this on AJAX Requests
+		// Don't trigger this on AJAX Requests.
 		if ( defined( 'DOING_AJAX' ) && DOING_AJAX ) {
```
Comment (mikejolley) [inline includes/class-wc-tracker.php]: Why is this needed? CRON is not an AJAX request and thats the only time this is called right?

File: includes/class-wc-tracker.php [hunk 8]
```diff
@@ -56,25 +60,25 @@ public static function send_tracking_data( $override = false ) {
...
-				'body'        => json_encode( $params ),
-				'cookies'     => array(),
-			)
-		);
+			'method'      => 'POST',
+			'timeout'     => 45,
+			'redirection' => 5,
+			'httpversion' => '1.0',
+			'blocking'    => false,
+			'headers'     => array( 'user-agent' => 'WooCommerceTracker/' . md5( esc_url( home_url( '/' ) ) ) . ';' ),
```
Comment (mikejolley) [inline includes/class-wc-tracker.php]: `esc_url_raw` more appropriate.

File: includes/class-wc-tracker.php [hunk 9]
```diff
@@ -56,25 +60,25 @@ public static function send_tracking_data( $override = false ) {
...
+			'body'        => json_encode( $params ),
```
Comment (mikejolley) [inline includes/class-wc-tracker.php]: This should use `wp_json_encode`

File: includes/class-wc-tracker.php [hunk 10]
```diff
@@ -83,71 +87,79 @@ private static function get_last_send_time() {
...
 
 	/**
 	 * Get the current theme info, theme name and version.
+	 *
 	 * @return array
 	 */
 	public static function get_theme_info() {
 		$theme_data        = wp_get_theme();
 		$theme_child_theme = is_child_theme() ? 'Yes' : 'No';
 		$theme_wc_support  = ! current_theme_supports( 'woocommerce' ) ? 'No' : 'Yes';
```
Comment (mikejolley) [inline includes/class-wc-tracker.php]: Use `wc_bool_to_string`

File: includes/class-wc-tracker.php [hunk 11]
```diff
@@ -171,13 +183,14 @@ private static function get_wordpress_info() {
 
 	/**
 	 * Get server related info.
+	 *
 	 * @return array
 	 */
 	private static function get_server_info() {
 		$server_data = array();
 
 		if ( isset( $_SERVER['SERVER_SOFTWARE'] ) && ! empty( $_SERVER['SERVER_SOFTWARE'] ) ) {
```
Comment (mikejolley) [inline includes/class-wc-tracker.php]: When using `empty` we don't need to check `isset`.

**Review by mikejolley — CHANGES_REQUESTED** [review]
@kloon Just made some final nitpicks.

My only other comment; when we call this manually during the setup wizard we might slow down the request. Could we schedule a single event for time() + 10 instead? Then the tracker request will be sent almost immediately but won't block the page load.

Thats all :)

File: includes/class-wc-tracker.php (hunk 7 above)
Comment (kloon) [reply includes/class-wc-tracker.php]: It was added because it was firing multiple times without this, think it has to do with the initial send which is not done via cron.

**Review by mikejolley — APPROVED** [review]
